diff -Naur bluez.orig/audio/avdtp.c bluez/audio/avdtp.c
--- bluez.orig/audio/avdtp.c	2013-10-22 09:36:17.622336701 +0300
+++ bluez/audio/avdtp.c	2013-10-22 13:44:21.245716272 +0300
@@ -1113,6 +1113,15 @@
 
 	if (state == AVDTP_STATE_IDLE &&
 				g_slist_find(session->streams, stream)) {
+		DBG("Removing stream %p. ", stream);
+
+		/* Quick hack to deal with abort requests made by stream callbacks */
+		if (session->req && session->req->stream == stream) {
+			DBG("Resetting stream to NULL for request %p (session %p)",
+				session->req, session);
+			session->req->stream = NULL;
+		}
+
 		session->streams = g_slist_remove(session->streams, stream);
 		stream_free(stream);
 	}
