--- old/audio/audio.conf	2011-07-11 15:42:10.995638292 +0800
+++ new/audio/audio.conf	2011-07-11 15:42:43.869855289 +0800
@@ -3,7 +3,7 @@
 # This section contains options which are not specific to any
 # particular interface
 [General]
-
+Enable=Gateway
 # Switch to master role for incoming connections (defaults to true)
 #Master=true

--- old/Makefile.am	2011-07-11 14:58:11.468606842 +0800
+++ new/Makefile.am	2011-07-11 15:01:30.835858646 +0800
@@ -277,7 +277,7 @@
 man_MANS = src/bluetoothd.8
 
 if DATAFILES
-conf_DATA += src/main.conf
+conf_DATA += src/main.conf audio/audio.conf
 endif
 
 EXTRA_DIST += src/genbuiltin src/bluetooth.conf \
 
