diff -Naur bluez.orig/plugins/hciops.c bluez/plugins/hciops.c
--- bluez.orig/plugins/hciops.c	2014-03-04 11:21:16.765990369 +0200
+++ bluez/plugins/hciops.c	2014-03-04 11:39:00.238029870 +0200
@@ -702,6 +702,7 @@
 		return -errno;
 
 	memcpy(dev->name, cp.name, 248);
+	dev->name[248] = '\0';
 	update_ext_inquiry_response(index);
 
 	return 0;
@@ -1649,9 +1650,10 @@
 		return;
 
 	memcpy(dev->name, rp->name, 248);
+	dev->name[248] = '\0';
 
 	if (!dev->pending) {
-		update_name(index, (char *) rp->name);
+		update_name(index, dev->name);
 		return;
 	}
 
