diff -Naur bluez.orig/audio/telephony-ofono.c bluez/audio/telephony-ofono.c
--- bluez.orig/audio/telephony-ofono.c	2013-10-23 11:24:30.272434561 +0300
+++ bluez/audio/telephony-ofono.c	2013-10-23 18:15:43.819309051 +0300
@@ -449,7 +449,13 @@
 		return;
 	}
 
-	/* Block memory dialing here; more proper would be to wait for
+	/* Fake memory dialing for memory location one. */
+	if (!strcmp(number, ">1")) {
+		telephony_last_dialed_number_req(telephony_device);
+		return;
+	}
+
+	/* Block other memory dialing; more proper would be to wait for
 	   ofono D-Bus reply, but I think that'd need audio device
 	   refcounting so that it doesn't potentially disappear while
 	   we wait for D-Bus reply. */
