diff -Naur bluez.orig/audio/telephony-ofono.c bluez/audio/telephony-ofono.c
--- bluez.orig/audio/telephony-ofono.c	2013-11-05 14:39:51.697539896 +0800
+++ bluez/audio/telephony-ofono.c	2013-11-05 14:41:18.405536401 +0800
@@ -54,7 +54,7 @@
 	char *number;
 	guint watch;
 
-	gboolean hold_status_pending;
+	gboolean status_pending;
 	gboolean waiting_for_answer;
 };
 
@@ -111,6 +111,8 @@
 	{ NULL }
 };
 
+static void update_call_status(void);
+
 static void waiting_for_answer_clear(struct voice_call *vc)
 {
 	DBG("");
@@ -129,28 +131,28 @@
 	return vc->waiting_for_answer;
 }
 
-static void hold_status_clear(struct voice_call *vc)
+static void status_clear(struct voice_call *vc)
 {
-	vc->hold_status_pending = FALSE;
+	vc->status_pending = FALSE;
 }
 
-static void hold_status_set_all(void)
+static void status_set_all(void)
 {
 	GSList *l;
 
 	for (l = calls; l != NULL; l = l->next) {
 		struct voice_call *vc = l->data;
-		vc->hold_status_pending = TRUE;
+		vc->status_pending = TRUE;
 	}
 }
 
-static gboolean hold_status_pending(void)
+static gboolean status_pending(void)
 {
 	GSList *l;
 
 	for (l = calls; l != NULL; l = l->next) {
 		struct voice_call *vc = l->data;
-		if (vc->hold_status_pending == TRUE)
+		if (vc->status_pending == TRUE)
 			return TRUE;
 	}
 
@@ -769,8 +771,8 @@
 	}
 
 done:
-	if (cme_err == CME_ERROR_NONE) /* wait for all hold statuses now */
-		hold_status_set_all();
+	if (cme_err == CME_ERROR_NONE) /* wait for all call statuses now */
+		status_set_all();
 
 	telephony_call_hold_rsp(telephony_device, cme_err);
 }
@@ -850,10 +852,8 @@
 
 	DBG("%s", vc->obj_path);
 
-	if (vc->status == CALL_STATUS_ACTIVE)
-		telephony_update_indicator(ofono_indicators, "call",
-							EV_CALL_INACTIVE);
-	else
+	update_call_status();
+	if (vc->status != CALL_STATUS_ACTIVE)
 		telephony_update_indicator(ofono_indicators, "callsetup",
 							EV_CALLSETUP_INACTIVE);
 
@@ -871,8 +871,8 @@
 {
 	DBG("");
 
-	if (hold_status_pending()) {
-		DBG("Hold statuses pending. ");
+	if (status_pending()) {
+		DBG("Call statuses pending. ");
 		return;
 	}
 
@@ -892,6 +892,26 @@
 	}
 }
 
+static void update_call_status(void)
+{
+	DBG("");
+
+	if (status_pending()) {
+		DBG("Call statuses pending. ");
+		return;
+	}
+
+	if (find_vc_with_status(CALL_STATUS_ACTIVE)) {
+		telephony_update_indicator(ofono_indicators,
+					"call",
+					EV_CALL_ACTIVE);
+	} else {
+		telephony_update_indicator(ofono_indicators,
+					"call",
+					EV_CALL_INACTIVE);
+	}
+}
+
 static gboolean handle_vc_property_changed(DBusConnection *conn,
 					DBusMessage *msg, void *data)
 {
@@ -917,7 +937,7 @@
 	if (g_str_equal(property, "State")) {
 		dbus_message_iter_get_basic(&sub, &state);
 		DBG("State %s", state);
-		hold_status_clear(vc);
+		status_clear(vc);
 		if (g_str_equal(state, "disconnected")) {
 			calls = g_slist_remove(calls, vc);
 			call_free(vc);
@@ -925,14 +945,13 @@
 			   of release and answer (AT+CHLD=1) processing */
 			answer_waiting_call();
 		} else if (g_str_equal(state, "active")) {
-			telephony_update_indicator(ofono_indicators,
-							"call", EV_CALL_ACTIVE);
-			telephony_update_indicator(ofono_indicators,
-							"callsetup",
-							EV_CALLSETUP_INACTIVE);
 			if (vc->status == CALL_STATUS_INCOMING)
 				telephony_calling_stopped_ind();
 			vc->status = CALL_STATUS_ACTIVE;
+			update_call_status();
+			telephony_update_indicator(ofono_indicators,
+							"callsetup",
+							EV_CALLSETUP_INACTIVE);
 		} else if (g_str_equal(state, "alerting")) {
 			telephony_update_indicator(ofono_indicators,
 					"callsetup", EV_CALLSETUP_ALERTING);
