diff -Naur bluez.orig/audio/telephony-ofono.c bluez/audio/telephony-ofono.c
--- bluez.orig/audio/telephony-ofono.c	2013-10-08 15:34:55.515396185 +0300
+++ bluez/audio/telephony-ofono.c	2013-10-08 15:44:58.943375584 +0300
@@ -933,7 +933,7 @@
 static void handle_network_property(const char *property, DBusMessageIter *variant)
 {
 	const char *status, *operator;
-	unsigned int signals_bar;
+	DBusBasicValue signals_bar;
 
 	if (g_str_equal(property, "Status")) {
 		dbus_message_iter_get_basic(variant, &status);
@@ -962,12 +962,12 @@
 		DBG("Operator is %s", operator);
 		g_free(net.operator_name);
 		net.operator_name = g_strdup(operator);
-	} else if (g_str_equal(property, "SignalStrength")) {
+	} else if (g_str_equal(property, "Strength")) {
 		dbus_message_iter_get_basic(variant, &signals_bar);
-		DBG("SignalStrength is %d", signals_bar);
-		net.signals_bar = signals_bar;
+		DBG("Strength is %u", (uint32_t)signals_bar.byt);
+		net.signals_bar = (uint32_t)signals_bar.byt;
 		telephony_update_indicator(ofono_indicators, "signal",
-						(signals_bar + 20) / 21);
+						(net.signals_bar + 20) / 21);
 	}
 }
 
